src/Math/Permutations.dfy

```diff
@@ -0,0 +1,50 @@
+/*******************************************************************************
+ *  Copyright by the contributors to the Dafny Project
+ *  SPDX-License-Identifier: MIT
+ *******************************************************************************/
+
+module Permutations {
+  import NatArith
+
+  /************
+   Definitions
+  ************/
+
+  function NumberOfPermutationsOf<T(==)>(s: seq<T>): nat {
+    |CalculateAllPermutationsOf(s)|
+  }
+
+  function CalculateAllPermutationsOf<T(==)>(s: seq<T>): (x: set<seq<T>>)
```

Comment:
Why compute them explicitely? We know what the number of permutation should be, and we could define a predicate to test if one sequence is a permutation of another.

Reply:
One can show (see https://dafny.org/blog/2023/11/08/cracking-the-coding-interview-in-dafny-permutations/) that the (finite) set computed by above function is equivalent to the (infinite) set
```
ghost function AllPermutationsOf<T(!new)>(s: seq<T>): iset<seq<T>> {
  iset p | multiset(p) == multiset(s)
}
```
For simplicity, I have excluded the material in this PR. I am planning to re-add it for correctness.